File: Project 1 - Bingo Game - I/bingo.py
import random

# Task 1 - Create the entry file and print welcome message
# Step 1: Entry File Setup
# Step 2: Welcome Banner Output
def print_welcome():
    print("Welcome to Bingo!")


# Task 2 - Define the 12 winning lines and FREE center index
# Step 1: Center Index Constant
FREE_INDEX = 12

# Step 2: Winning Lines Matrix
WIN_LINES = [
    (0, 1, 2, 3, 4), (5, 6, 7, 8, 9), (10, 11, 12, 13, 14),        # rows
    (15, 16, 17, 18, 19), (20, 21, 22, 23, 24),
    (0, 5, 10, 15, 20), (1, 6, 11, 16, 21), (2, 7, 12, 17, 22),    # columns
    (3, 8, 13, 18, 23), (4, 9, 14, 19, 24),
    (0, 6, 12, 18, 24), (4, 8, 12, 16, 20),                        # diagonals
]


# Task 3 - Create the Card class and generate numbers
class Card:
    def __init__(self):
        # Step 1: State Initialization
        self.numbers = self.generate()
        self.marked = [False] * 25
        self.marked[FREE_INDEX] = True

    def generate(self):
        cells = [None] * 25
        low = [1, 16, 31, 46, 61]
        high = [15, 30, 45, 60, 75]

        # Step 2: Column-Wise Number Generation
        for col in range(5):
            col_numbers = random.sample(range(low[col], high[col] + 1), 5)
            # Step 3: 1D Grid Coordinate Mapping & Center FREE Placement
            for row in range(5):
                cells[row * 5 + col] = col_numbers[row]
        cells[FREE_INDEX] = "FREE"
        return cells

    # Task 4 - Display the card
    def display(self, title="Your Card"):
        # Step 1: Header Title & Row Iteration
        print(f"\n{title}")
        for row in range(5):
            cells_in_row = []
            # Step 2: Cell Formatting & Daub Indicator
            for col in range(5):
                idx = row * 5 + col
                value = self.numbers[idx]
                text = str(value)
                if self.marked[idx]:
                    text = f"*{text}*"
                cells_in_row.append(text.center(6))
            # Step 3: Pipe Separators & Horizontal Dividers
            print("|".join(cells_in_row))
            if row != 4:
                print("-" * 34)

    # Task 5 - Implement mark(number)
    def mark(self, number):
        # Step 1: Linear Search for Target Value
        for i in range(25):
            # Step 2: Daub & Confirmation Return
            if self.numbers[i] == number and not self.marked[i]:
                self.marked[i] = True
                return True
        return False

    # Task 6 - Implement checkWin()
    def check_win(self):
        # Step 1: Iterate Winning Line Definitions
        for line in WIN_LINES:
            # Step 2: Line Verification & Early Exit
            line_complete = True
            for idx in line:
                if not self.marked[idx]:
                    line_complete = False
                    break
            if line_complete:
                return True
        return False


# Task 7 - Create the Player class
class Player:
    def __init__(self, raw_name):
        # Step 1: Name Sanitization & Default Assignment
        if raw_name and raw_name.strip():
            self.name = raw_name.strip()
        else:
            self.name = "Player"
        # Step 2: Card Ownership & Victory Flag
        self.card = Card()
        self.has_won = False


# Task 8 - Build a list of players
def build_players():
    # Step 1: Player Count Prompt & Defensive Validation
    while True:
        user_input = input("How many players (1-6)? ").strip()
        if user_input.isdigit() and 1 <= int(user_input) <= 6:
            num_players = int(user_input)
            break
        print("Please enter a number from 1 to 6.")

    # Step 2: Roster Instantiation Loop
    players = []
    for i in range(1, num_players + 1):
        name = input(f"Enter name for Player {i}: ")
        players.append(Player(name))
    return players


# Task 9 - Create the Game class skeleton
class Game:
    def __init__(self, players):
        # Step 1: State Binding
        self.players = players
        self.number_pool = []
        self.called_numbers = []
        # Step 2: Number Pool Generation & Shuffling
        self.reset_number_pool()

    def reset_number_pool(self):
        self.number_pool = list(range(1, 76))
        random.shuffle(self.number_pool)
        self.called_numbers = []

    # Task 10 - Implement drawNumber()
    def draw_number(self):
        # Step 1: Draw from Pool
        number = self.number_pool.pop()
        # Step 2: History Logging & Return
        self.called_numbers.append(number)
        return number

    # Task 11 - Broadcast a drawn number to every player
    def broadcast(self, number):
        # Step 1: Roster Broadcast Loop
        for player in self.players:
            # Step 2: Console Daub Notification
            if player.card.mark(number):
                print(f"  -> marked on {player.name}'s card")

    # Task 12 - Check every player for a win, every round
    def check_winners(self):
        # Step 1: Winner Collection Scan
        winners = []
        for player in self.players:
            # Step 2: Flag State & Winner Accumulation
            if player.card.check_win():
                player.has_won = True
                winners.append(player)
        return winners

    # Task 13 - Write the interactive play() loop
    def play(self):
        # Step 1: Match Setup & Card Reveal
        self.reset_number_pool()
        for player in self.players:
            player.has_won = False
            player.card = Card()
            input(f"\n{player.name}, press Enter to draw your card:")
            player.card.display(f"{player.name}'s Card")

        # Step 2: Match Start Indicator & Status Tracking
        print("\n===== MATCH STARTED =====")
        winners = []
        turn = 0

        # Step 3: Turn-Based Draw Loop & Interactive Pacing
        while self.number_pool and not winners:
            active_player = self.players[turn % len(self.players)]
            input(f"\n{active_player.name}, press Enter to draw the next number:")
            number = self.draw_number()
            print(f"\nNumber called: {number}  (draw #{len(self.called_numbers)})")

            # Step 4: Broadcast, Win Check & Turn Increment
            self.broadcast(number)
            winners = self.check_winners()
            turn += 1

        # Step 5: Match Conclusion Dispatch
        self.announce_result(winners)

    # Task 14 - Announce the result
    def announce_result(self, winners):
        # Step 1: Branch on Win vs. Draw
        if winners:
            # Step 2: Winner Showcase & Final Card Grid Display
            for player in self.players:
                player.card.display(f"{player.name}'s Final Card")
            names = ", ".join(player.name for player in winners)
            print(f"\nBINGO! {names} won after {len(self.called_numbers)} numbers called!")
        else:
            # Step 3: Exhausted Pool Draw Message
            print("\nAll 75 numbers were called with no winner. It's a draw!")

    # Task 15 - Implement askReplay()
    @staticmethod
    def ask_replay():
        # Step 1: Interactive Loop & Choice Prompt
        while True:
            answer = input("Play again? (y/n): ").strip().lower()
            # Step 2: Sanitization & Normalized Matching
            if answer == "y":
                return True
            if answer == "n":
                return False
            print("Please type y or n.")

    # Task 16 - Implement playMany()
    def play_many(self):
        # Step 1: Multi-Match Indefinite Loop
        while True:
            self.play()
            # Step 2: Rematch Evaluation & Exit Farewell
            if not self.ask_replay():
                print("Thanks for playing Bingo!")
                break


# Task 17 - Assemble the main() entry point
def main():
    # Step 1: Component Initialization Pipeline
    print_welcome()
    players = build_players()
    game = Game(players)
    # Step 2: Kickoff Application Session
    game.play_many()


if __name__ == "__main__":
    main()

File: Project 1 - Bingo Game - I/bingo.cpp
#include <algorithm>
#include <iostream>
#include <random>
#include <string>
#include <vector>

using namespace std;

// Task 1 - Create the entry file and print welcome message
// Step 1: Entry File Setup
// Step 2: Welcome Banner Output
void printWelcome() {
    cout << "Welcome to Bingo!\n";
}

// Task 2 - Define the 12 winning lines and FREE center index
// Step 1: Center Index Constant
const int FREE_INDEX = 12;

// Step 2: Winning Lines Matrix
static const vector<vector<int>> WIN_LINES = {
    {0, 1, 2, 3, 4}, {5, 6, 7, 8, 9}, {10, 11, 12, 13, 14},        // rows
    {15, 16, 17, 18, 19}, {20, 21, 22, 23, 24},
    {0, 5, 10, 15, 20}, {1, 6, 11, 16, 21}, {2, 7, 12, 17, 22},    // columns
    {3, 8, 13, 18, 23}, {4, 9, 14, 19, 24},
    {0, 6, 12, 18, 24}, {4, 8, 12, 16, 20},                        // diagonals
};

// Random number engine for shuffling
static mt19937 rng(random_device{}());

// Helper function to center text within a given column width
string center(const string& text, int width) {
    int pad = width - static_cast<int>(text.size());
    int left = pad / 2;
    int right = pad - left;
    return string(left > 0 ? left : 0, ' ') + text + string(right > 0 ? right : 0, ' ');
}

// Task 3 - Create the Card class and generate numbers
class Card {
public:
    vector<int> numbers;   // -1 marks the FREE cell
    vector<bool> marked;

    Card() : numbers(25), marked(25, false) {
        // Step 1: State Initialization
        generate();
        marked[FREE_INDEX] = true;
    }

    void generate() {
        int low[5]  = {1, 16, 31, 46, 61};
        int high[5] = {15, 30, 45, 60, 75};

        // Step 2: Column-Wise Number Generation
        for (int col = 0; col < 5; ++col) {
            vector<int> colNumbers;
            for (int v = low[col]; v <= high[col]; ++v) {
                colNumbers.push_back(v);
            }
            shuffle(colNumbers.begin(), colNumbers.end(), rng);
            // Step 3: 1D Grid Coordinate Mapping & Center FREE Placement
            for (int row = 0; row < 5; ++row) {
                numbers[row * 5 + col] = colNumbers[row];
            }
        }
        numbers[FREE_INDEX] = -1;
    }

    // Task 4 - Display the card
    void display(const string& title) const {
        // Step 1: Header Title & Row Iteration
        cout << "\n" << title << "\n";
        for (int row = 0; row < 5; ++row) {
            string line = "";
            // Step 2: Cell Formatting & Daub Indicator
            for (int col = 0; col < 5; ++col) {
                int idx = row * 5 + col;
                string text = (numbers[idx] == -1) ? "FREE" : to_string(numbers[idx]);
                if (marked[idx]) text = "*" + text + "*";
                line += center(text, 6);
                if (col != 4) line += "|";
            }
            // Step 3: Pipe Separators & Horizontal Dividers
            cout << line << "\n";
            if (row != 4) cout << string(34, '-') << "\n";
        }
    }

    // Task 5 - Implement mark(number)
    bool mark(int number) {
        // Step 1: Linear Search for Target Value
        for (int i = 0; i < 25; ++i) {
            // Step 2: Daub & Confirmation Return
            if (numbers[i] == number && !marked[i]) {
                marked[i] = true;
                return true;
            }
        }
        return false;
    }

    // Task 6 - Implement checkWin()
    bool checkWin() const {
        // Step 1: Iterate Winning Line Definitions
        for (const auto& line : WIN_LINES) {
            // Step 2: Line Verification & Early Exit
            bool lineComplete = true;
            for (int idx : line) {
                if (!marked[idx]) {
                    lineComplete = false;
                    break;
                }
            }
            if (lineComplete) return true;
        }
        return false;
    }
};

// Task 7 - Create the Player class
class Player {
public:
    string name;
    Card card;
    bool hasWon = false;

    explicit Player(const string& rawName) {
        // Step 1: Name Sanitization & Default Assignment
        size_t start = 0;
        while (start < rawName.size() && (rawName[start] == ' ' || rawName[start] == '\t')) {
            start++;
        }
        if (start == rawName.size()) {
            name = "Player";
        } else {
            name = rawName.substr(start);
        }
        // Step 2: Card Ownership & Victory Flag
        card = Card();
        hasWon = false;
    }
};

// Task 8 - Build a list of players
vector<Player> buildPlayers() {
    // Step 1: Player Count Prompt & Defensive Validation
    int numPlayers = 0;
    while (true) {
        cout << "How many players (1-6)? ";
        string userInput;
        getline(cin, userInput);
        bool allDigits = !userInput.empty();
        for (char c : userInput) {
            if (!isdigit(c)) {
                allDigits = false;
                break;
            }
        }
        if (allDigits) {
            numPlayers = stoi(userInput);
            if (numPlayers >= 1 && numPlayers <= 6) break;
        }
        cout << "Please enter a number from 1 to 6.\n";
    }

    // Step 2: Roster Instantiation Loop
    vector<Player> players;
    for (int i = 1; i <= numPlayers; ++i) {
        cout << "Enter name for Player " << i << ": ";
        string name;
        getline(cin, name);
        players.push_back(Player(name));
    }
    return players;
}

// Task 9 - Create the Game class skeleton
class Game {
public:
    vector<Player>& players;
    vector<int> numberPool;
    vector<int> calledNumbers;

    explicit Game(vector<Player>& players) : players(players) {
        // Step 1: State Binding
        // Step 2: Number Pool Generation & Shuffling
        resetNumberPool();
    }

    void resetNumberPool() {
        numberPool.clear();
        for (int v = 1; v <= 75; ++v) {
            numberPool.push_back(v);
        }
        shuffle(numberPool.begin(), numberPool.end(), rng);
        calledNumbers.clear();
    }

    // Task 10 - Implement drawNumber()
    int drawNumber() {
        // Step 1: Draw from Pool
        int number = numberPool.back();
        numberPool.pop_back();
        // Step 2: History Logging & Return
        calledNumbers.push_back(number);
        return number;
    }

    // Task 11 - Broadcast a drawn number to every player
    void broadcast(int number) {
        // Step 1: Roster Broadcast Loop
        for (auto& player : players) {
            // Step 2: Console Daub Notification
            if (player.card.mark(number)) {
                cout << "  -> marked on " << player.name << "'s card\n";
            }
        }
    }

    // Task 12 - Check every player for a win, every round
    vector<string> checkWinners() {
        // Step 1: Winner Collection Scan
        vector<string> winners;
        for (auto& player : players) {
            // Step 2: Flag State & Winner Accumulation
            if (player.card.checkWin()) {
                player.hasWon = true;
                winners.push_back(player.name);
            }
        }
        return winners;
    }

    // Task 13 - Write the interactive play() loop
    void play() {
        // Step 1: Match Setup & Card Reveal
        resetNumberPool();
        for (auto& player : players) {
            player.hasWon = false;
            player.card = Card();
            cout << "\n" << player.name << ", press Enter to draw your card:";
            string enterKey;
            getline(cin, enterKey);
            player.card.display(player.name + "'s Card");
        }

        // Step 2: Match Start Indicator & Status Tracking
        cout << "\n===== MATCH STARTED =====\n";
        vector<string> winners;
        size_t turn = 0;

        // Step 3: Turn-Based Draw Loop & Interactive Pacing
        while (!numberPool.empty() && winners.empty()) {
            Player& activePlayer = players[turn % players.size()];
            cout << "\n" << activePlayer.name << ", press Enter to draw the next number:";
            string enterKey;
            getline(cin, enterKey);

            int number = drawNumber();
            cout << "\nNumber called: " << number
                 << "  (draw #" << calledNumbers.size() << ")\n";

            // Step 4: Broadcast, Win Check & Turn Increment
            broadcast(number);
            winners = checkWinners();
            turn++;
        }

        // Step 5: Match Conclusion Dispatch
        announceResult(winners);
    }

    // Task 14 - Announce the result
    void announceResult(const vector<string>& winners) {
        // Step 1: Branch on Win vs. Draw
        if (!winners.empty()) {
            // Step 2: Winner Showcase & Final Card Grid Display
            for (auto& player : players) {
                player.card.display(player.name + "'s Final Card");
            }
            cout << "\nBINGO! ";
            for (size_t i = 0; i < winners.size(); ++i) {
                cout << winners[i];
                if (i + 1 != winners.size()) cout << ", ";
            }
            cout << " won after " << calledNumbers.size()
                 << " numbers called!\n";
        } else {
            // Step 3: Exhausted Pool Draw Message
            cout << "\nAll 75 numbers were called with no winner. It's a draw!\n";
        }
    }

    // Task 15 - Implement askReplay()
    static bool askReplay() {
        // Step 1: Interactive Loop & Choice Prompt
        while (true) {
            cout << "Play again? (y/n): ";
            string answer;
            getline(cin, answer);
            // Step 2: Sanitization & Normalized Matching
            if (answer == "y") return true;
            if (answer == "n") return false;
            cout << "Please type y or n.\n";
        }
    }

    // Task 16 - Implement playMany()
    void playMany() {
        // Step 1: Multi-Match Indefinite Loop
        while (true) {
            play();
            // Step 2: Rematch Evaluation & Exit Farewell
            if (!askReplay()) {
                cout << "Thanks for playing Bingo!\n";
                break;
            }
        }
    }
};

// Task 17 - Assemble the main() entry point
int main() {
    // Step 1: Component Initialization Pipeline
    printWelcome();
    vector<Player> players = buildPlayers();
    Game game(players);
    // Step 2: Kickoff Application Session
    game.playMany();
    return 0;
}